File: spherepop-kernel/src/history.rs
//! Histories, the option space, and deterministic replay.
//!
//! `History` is the free monoid over the event alphabet (append, `++`,
//! empty). `apply` is the per-event transition function whose fold over
//! a history is `Replay`. `apply` is deliberately a pure function of
//! `(State, Event)` alone — no wall-clock reads, no ambient globals — so
//! that Invariant (Deterministic Replay) is checkable rather than merely
//! asserted: replaying the same `History` twice is guaranteed by
//! construction to produce equal `State` values.

use crate::event::{Event, EventKind, LogPos, ObjectId, RuleId};
use std::collections::HashSet;

#[derive(Default, Clone, Debug, PartialEq, Eq)]
pub struct State {
    /// Omega: the current option space.
    pub option_space: HashSet<ObjectId>,
    /// Symbols already committed via Pop.
    pub committed: HashSet<ObjectId>,
    /// Bind facts: (a, b, tag).
    pub bound: HashSet<(ObjectId, ObjectId, String)>,
    /// (position, target, reason) for every Refuse event, target=0 for
    /// refusals that target a Bind pair rather than a bare object (see
    /// Unlink-sugar, which is Refuse(Bind(a,b))).
    pub refused: Vec<(LogPos, Option<ObjectId>, String)>,
    /// Audit trail of Collapse invocations: (position, rule). Deliberately
    /// does NOT store the observed value c(H) itself — see Requirement
    /// (Observation Non-Interference) in the specification.
    pub observed: Vec<(LogPos, RuleId)>,
}

#[derive(Default, Clone, Debug, PartialEq, Eq)]
pub struct History {
    events: Vec<Event>,
}

impl History {
    pub fn new() -> Self {
        History::default()
    }

    pub fn len(&self) -> usize {
        self.events.len()
    }

    pub fn is_empty(&self) -> bool {
        self.events.is_empty()
    }

    pub fn as_slice(&self) -> &[Event] {
        &self.events
    }

    /// The only mutator: append. There is no remove/undo, matching the
    /// free-monoid structure and the Irreversibility corollary.
    pub(crate) fn push(&mut self, event: Event) {
        self.events.push(event);
    }

    pub fn events_of_kind(&self, kind: EventKind) -> impl Iterator<Item = &Event> {
        self.events.iter().filter(move |e| e.kind == kind)
    }

    /// Meld: parallel composition of two independently-generated histories
    /// (the free monoidal tensor). Structural fact about the history
    /// monoid; NOT required by the concurrency clause of the Completeness
    /// Theorem, which goes through Bind alone. Provided here because it is
    /// still a true, useful operation for combining two logs.
    pub fn meld(&mut self, other: &History) {
        self.events.extend(other.events.iter().cloned());
    }

    /// Replay: fold `apply` over the history from a given initial Omega.
    pub fn replay(&self, omega_0: &HashSet<ObjectId>) -> State {
        let mut s = State {
            option_space: omega_0.clone(),
            ..State::default()
        };
        for e in &self.events {
            apply(&mut s, e);
        }
        s
    }
}

pub fn apply(s: &mut State, e: &Event) {
    match e.kind {
        EventKind::Pop => {
            let x = e.a.expect("Pop event missing target `a`");
            s.option_space.remove(&x); // |Omega| decreases by exactly one
            s.committed.insert(x);
        }
        EventKind::Refuse => {
            s.refused.push((e.pos, e.a, e.reason.clone().unwrap_or_default()));
            // Omega is untouched: refusal documents, it does not foreclose.
        }
        EventKind::Bind => {
            let a = e.a.expect("Bind event missing `a`");
            let b = e.b.expect("Bind event missing `b`");
            s.bound.insert((a, b, e.tag.clone().unwrap_or_default()));
        }
        EventKind::Collapse => {
            // Collapse never mutates committed/bound/refused, and its
            // observable value c(H) is computed separately (see
            // `collapse.rs`), never stored in State itself.
            let rule = e.rule.expect("Collapse event missing `rule`");
            s.observed.push((e.pos, rule));
        }
    }
}

/// The event-weight function w and the generalised possibility functional
/// Pi(H, Omega) = |Omega| + sum_e w(e), used by the Conservation Law test.
pub fn possibility_functional(state: &State) -> usize {
    let popped = state.committed.len(); // w(Pop) = 1, all other kinds w = 0
    state.option_space.len() + popped
}

File: spherepop-go/history.go
package spherepop

// Binding is a recorded dependency. Relatedness never implies identity.
type Binding struct {
	A   ObjectID
	B   ObjectID
	Tag string
}

type Refusal struct {
	Pos    LogPos
	Target *ObjectID
	Other  *ObjectID
	Reason string
}

type Observation struct {
	Pos  LogPos
	Rule RuleID
}

// State is a deterministic replay product, never the authoritative source.
type State struct {
	OptionSpace map[ObjectID]struct{}
	Committed   map[ObjectID]struct{}
	Bound       map[Binding]struct{}
	Refused     []Refusal
	Observed    []Observation
}

func NewState(omega []ObjectID) State {
	s := State{
		OptionSpace: make(map[ObjectID]struct{}, len(omega)),
		Committed:   make(map[ObjectID]struct{}),
		Bound:       make(map[Binding]struct{}),
	}
	for _, id := range omega {
		s.OptionSpace[id] = struct{}{}
	}
	return s
}

func (s State) Clone() State {
	out := NewState(nil)
	for id := range s.OptionSpace {
		out.OptionSpace[id] = struct{}{}
	}
	for id := range s.Committed {
		out.Committed[id] = struct{}{}
	}
	for b := range s.Bound {
		out.Bound[b] = struct{}{}
	}
	out.Refused = append([]Refusal(nil), s.Refused...)
	out.Observed = append([]Observation(nil), s.Observed...)
	return out
}

func Apply(s *State, e Event) {
	switch e.Kind {
	case Pop:
		delete(s.OptionSpace, *e.A)
		s.Committed[*e.A] = struct{}{}
	case Refuse:
		r := ""
		if e.Reason != nil {
			r = *e.Reason
		}
		s.Refused = append(s.Refused, Refusal{e.Pos, e.A, e.B, r})
	case Bind:
		t := ""
		if e.Tag != nil {
			t = *e.Tag
		}
		s.Bound[Binding{*e.A, *e.B, t}] = struct{}{}
	case Collapse:
		s.Observed = append(s.Observed, Observation{e.Pos, *e.Rule})
	}
}

// History is append-only. Events returns a defensive copy.
type History struct{ events []Event }

func NewHistory() History  { return History{} }
func (h History) Len() int { return len(h.events) }

func (h History) Events() []Event {
	return append([]Event(nil), h.events...)
}

func (h History) EventsOfKind(kind EventKind) []Event {
	out := make([]Event, 0)
	for _, e := range h.events {
		if e.Kind == kind {
			out = append(out, e)
		}
	}
	return out
}

func (h *History) append(e Event) { h.events = append(h.events, e) }

// Meld is structural composition of histories, not a fifth primitive.
func (h History) Meld(other History) History {
	out := NewHistory()
	out.events = append(out.events, h.events...)
	out.events = append(out.events, other.events...)
	for i := range out.events {
		out.events[i].Pos = LogPos(i)
	}
	return out
}

func (h History) Replay(omega []ObjectID) State {
	s := NewState(omega)
	for _, e := range h.events {
		Apply(&s, e)
	}
	return s
}

// PossibilityFunctional is |Omega| + committed Pop count. All non-Pop
// primitive weights are zero in SPEC v1.
func PossibilityFunctional(s State) int {
	return len(s.OptionSpace) + len(s.Committed)
}
